hummingbot/hummingbot
# Fix/time synchronizer and abandoned orders issue

**Before submitting this PR, please make sure**:


**A description of the changes proposed in the pull request**:
- Change in the TimeSynchronizer logic to estimate timestamps giving more value to resent samples than older samples.
- Refactor in ExchangePyBase to include in the abstract class the common logic about order HTTP status update loop. By default it always tries now to not only get updates for the orders, but also check if there have been any order fills (trades) impacting them. All subclasses of ExchangePyBase have been updated to implement the required logic to get the order fills and order updates.
- Added logic in ClientOrderTracker to keep track of all the orders that are marked as failed because they have not been found in the exchange or there were errors retrieving updates from the exchange for them more times than the specified limit. These orders are called "lost orders".
- Added new loop task in ExchangePyBase class to regularly try to cancel all lost orders. The loop also tries to get updates for the orders (to find out when they are no longer alive in the server) and also requests order fill updates for them. This is a fall back functionality to avoid leaving stale orders alive in the exchange.


**Tests performed by the developer**:
Added unit tests for all new functionalities.
Tested the changes manually by running a PMM strategy connected with Binance and operating with ETH-DAI


**Test summary**:
- Install the feature branch without issue
-  Start the client and setup a PMM strategy with multiple order levels using Binance, Bitmart, Gate_io, KuCoin, and OKX connector.
- Desynchronise  local machine time by adjust the time to a later time or advance from the current time.
-  Monitor the strategy creation and cancelation: orders are getting canceled and created without issue
- Run the test bot for several hours - initially a cancelation issue were reported and had been addressed.
- Run a long term test bot ( run for more than 2 days)
-  Review the logs: Binance still get server time error but it was greatly reduce and no hanging orders created.
- Other exchange had a cancelation error but due to an order getting filed which is expected


**Tips for QA testing**:
Test that new client instances started with the new version do not leave orders alive in the server after removing them in the bot list of active orders. Initially please test these with Binance.
We should also see less issues in Binance strategies execution related to request being rejected due to an invalid timestamp.
Please also run a quick validation that all connectors implemented as subclasses of ExchangePyBase still can be used without problems:
- Kucoin
- AscendEx
- OKX
- GateIO
- Bitmart
- Latoken


This PR comes from https://github.com/CoinAlpha/hummingbot/pull/276

PRP: https://snapshot.org/#/hbot-prp.eth/proposal/0xf14a04c901aaa411d10ab0b00cefda7b888692d36e8ef20afaa60aaeede489df
